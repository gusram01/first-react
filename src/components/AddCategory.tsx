import React, { ChangeEvent, Dispatch, FormEvent, SetStateAction, useState } from 'react'
import PropTypes from 'prop-types';

interface Props {
  setCategories: Dispatch<SetStateAction<string[]>>;
}

const AddCategory: React.FC<Props> = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState('');
  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };
  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories(category => [inputValue, ...category]);
      setInputValue('');
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange} />
    </ form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

export default AddCategory;