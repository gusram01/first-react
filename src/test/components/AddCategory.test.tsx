import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('AddCategory tests', () => {

  const setCategories = jest.fn();
  let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test('should render AddCategory correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('input text change value', () => {
    const input = wrapper.find('input');
    const value = 'Hello world';
    input.simulate('change', {
      target: { value }
    });
  });

  test('shold not post info on submit', () => {
    wrapper.find('form').simulate('submit', { preventDefault() { } });

    expect(setCategories).not.toHaveBeenCalled();
  });

  test('should call setCategories and clean input text', () => {
    const value = 'something';
    const input = wrapper.find('input');
    input.simulate('change', { target: { value } });
    wrapper.find('form').
      simulate('submit', { preventDefault() { } });
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
    expect(input.prop('value')).toBe('');
  });

});
