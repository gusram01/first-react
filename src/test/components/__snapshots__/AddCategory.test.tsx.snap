// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory tests should render AddCategory correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
